File: app/src/main/java/com/ricamgar/notify/reminderslist/RemindersListAdapter.java
package com.ricamgar.notify.reminderslist;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;


import com.jakewharton.rxbinding.view.RxView;
import com.ricamgar.notify.R;
import com.ricamgar.notify.domain.reminder.model.Reminder;

import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;

import rx.Observable;
import rx.functions.Action1;

public class RemindersListAdapter extends RecyclerView.Adapter<RemindersListAdapter.ReminderViewHolder> implements
        Action1<List<Reminder>> {

    private final ReminderListener listener;
    private final int mode;
    private final Context context;
    private List<Reminder> items = Collections.emptyList();

    public RemindersListAdapter(Context context, ReminderListener listener, int mode) {
        this.context = context;
        this.listener = listener;
        this.mode = mode;
    }

    @Override
    public ReminderViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
        View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.reminder_item, viewGroup, false);
        return new ReminderViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ReminderViewHolder reminderViewHolder, int position) {
        Reminder reminder = items.get(position);
        reminderViewHolder.description.setText(reminder.description);
        reminderViewHolder.enabled.setText(reminder.done ? "SHOW NOTIFICATION" : "MARK AS DONE");
        RxView.clicks(reminderViewHolder.enabled)
                .doOnNext(o1 -> {
                    items.remove(position);
                    notifyItemRemoved(position);
                })
                .delay(500, TimeUnit.MILLISECONDS)
                .subscribe(o -> listener.onReminderChecked(reminder, !reminder.done));

        reminderViewHolder.itemView.setOnClickListener(view -> {
            listener.onReminderClicked(reminder);
        });
        if (mode == 0) {
            reminderViewHolder.delete.setVisibility(View.GONE);
        } else {
            RxView.clicks(reminderViewHolder.delete)
                    .doOnNext(o1 -> {
                        items.remove(position);
                        notifyItemRemoved(position);
                    })
                    .delay(500, TimeUnit.MILLISECONDS)
                    .subscribe(o -> listener.onReminderDeleted(reminder));

            reminderViewHolder.delete.setVisibility(View.VISIBLE);
        }

    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    @Override
    public void call(List<Reminder> newItems) {
        this.items = newItems;
        notifyDataSetChanged();
    }

    final static class ReminderViewHolder extends RecyclerView.ViewHolder {

        public final TextView description;
        public final TextView enabled;
        public final View delete;

        public ReminderViewHolder(View itemView) {
            super(itemView);
            description = (TextView) itemView.findViewById(R.id.description);
            enabled = (TextView) itemView.findViewById(R.id.enable_btn);
            delete = itemView.findViewById(R.id.delete);
        }
    }

    public interface ReminderListener {
        void onReminderClicked(Reminder reminder);

        void onReminderChecked(Reminder reminder, boolean checked);

        void onReminderDeleted(Reminder reminder);
    }
}

File: app/src/main/java/com/ricamgar/notify/reminderslist/RemindersListAdapter.kt
package com.ricamgar.notify.reminderslist

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.jakewharton.rxbinding.view.RxView
import com.ricamgar.notify.R
import com.ricamgar.notify.domain.reminder.model.Reminder

import java.util.concurrent.TimeUnit

import rx.functions.Action1

class RemindersListAdapter(
  private val listener: ReminderListener,
  private val mode: Int
) : RecyclerView.Adapter<RemindersListAdapter.ReminderViewHolder>(), Action1<MutableList<Reminder>> {

  private var items: MutableList<Reminder> = mutableListOf()

  override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): ReminderViewHolder {
    val view = LayoutInflater.from(viewGroup.context).inflate(R.layout.reminder_item, viewGroup, false)
    return ReminderViewHolder(view)
  }

  override fun onBindViewHolder(reminderViewHolder: ReminderViewHolder, position: Int) {
    val reminder = items[position]
    reminderViewHolder.description.text = reminder.description
    reminderViewHolder.enabled.text = if (reminder.done) "SHOW NOTIFICATION" else "MARK AS DONE"
    RxView.clicks(reminderViewHolder.enabled)
      .doOnNext {
        items.removeAt(position)
        notifyItemRemoved(position)
      }
      .delay(500, TimeUnit.MILLISECONDS)
      .subscribe { listener.onReminderChecked(reminder, !reminder.done) }

    reminderViewHolder.itemView.setOnClickListener { listener.onReminderClicked(reminder) }
    if (mode == 0) {
      reminderViewHolder.delete.visibility = View.GONE
    } else {
      RxView.clicks(reminderViewHolder.delete)
        .doOnNext {
          items.removeAt(position)
          notifyItemRemoved(position)
        }
        .delay(500, TimeUnit.MILLISECONDS)
        .subscribe { listener.onReminderDeleted(reminder) }

      reminderViewHolder.delete.visibility = View.VISIBLE
    }

  }

  override fun getItemCount(): Int = items.size

  override fun call(newItems: MutableList<Reminder>) {
    this.items = newItems
    notifyDataSetChanged()
  }

  class ReminderViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
    val description: TextView = itemView.findViewById<View>(R.id.description) as TextView
    val enabled: TextView = itemView.findViewById<View>(R.id.enable_btn) as TextView
    val delete: View = itemView.findViewById(R.id.delete)
  }

  interface ReminderListener {
    fun onReminderClicked(reminder: Reminder)

    fun onReminderChecked(reminder: Reminder, checked: Boolean)

    fun onReminderDeleted(reminder: Reminder)
  }
}
